File: token/token.go
package token

type TokenType string

type Token struct
{
    Type TokenType
    Literal string
}

const (
    ILLEGAL   = "ILLEGAL"
    EOF       = "EOF"

    // Identifiers & literals
    IDENT     = "IDENT"
    INT       = "INT"

    // Operators
    ASSIGN    = "="
    PLUS      = "+"
    MINUS     = "-"
    SLASH     = "/"
    ASTERISK  = "*"
    LT        = "<"
    GT        = ">"
    BANG      = "!"
    EQ        = "=="
    NEQ       = "!="

    // Delimiters
    COMMA     = ","
    SEMICOLON = ";"
    LPAREN    = "("
    RPAREN    = ")"
    LBRACE    = "{"
    RBRACE    = "}"

    // Keywords
    FUNCTION  = "FUNCTION"
    LET       = "LET"
    RETURN    = "RETURN"
    IF        = "IF"
    ELSE      = "ELSE"
    TRUE      = "TRUE"
    FALSE     = "FALSE"
)

var keywords = map[string]TokenType {
    "fn"    : FUNCTION,
    "let"   : LET,
    "return": RETURN,
    "true"  : TRUE,
    "false" : FALSE,
    "if"    : IF,
    "else"  : ELSE,
}

func LookupIdent(ident string) TokenType {
    if tok, ok := keywords[ident]; ok {
	return tok
    }
    return IDENT
}

File: src/lexer/token.rs
#[derive(Eq, PartialEq, Hash, Clone, Copy, Debug)]
pub enum TokenType {
    ILLEGAL = 0,
    EOF,
    // Idetifier & literals
    IDENT,
    INT,
    //Operators
    ASSIGN,
    PLUS,
    MINUS,
    SLASH,
    ASTERISK,
    LT,
    GT,
    BANG,
    EQ,
    NEQ,

    // Delimiters
    COMMA,
    SEMICOLON,
    LPAREN,
    RPAREN,
    LBRACE,
    RBRACE,

    // Keywords
    FUNCTION,
    LET,
    RETURN,
    IF,
    ELSE,
    TRUE,
    FALSE,
}

// const TOKEN_LITERAL: HashMap<TokenType, String> = [
//     (TokenType::ILLEGAL, "ILLEGAL"),
//     (TokenType::EOF, "EOF"),
//     (TokenType::IDENT, "IDENT"),
//     (TokenType::INT, "INT"),
//     (TokenType::ASSIGN, "="),
//     (TokenType::PLUS, "+"),
//     (TokenType::MINUS, "-"),
//     (TokenType::SLASH, "/"),
//     (TokenType::ASTERISK, "*"),
//     (TokenType::LT, "<"),
//     (TokenType::GT, ">"),
//     (TokenType::BANG, "!"),
//     (TokenType::EQ, "=="),
//     (TokenType::NEQ, "!="),
//     (TokenType::COMMA, ","),
//     (TokenType::SEMICOLON, ";"),
//     (TokenType::LPAREN, "("),
//     (TokenType::RPAREN, ")"),
//     (TokenType::LBRACE, "{"),
//     (TokenType::RBRACE, "}"),
//     (TokenType::FUNCTION, "FUNCTION"),
//     (TokenType::LET, "LET"),
//     (TokenType::RETURN, "RETURN"),
//     (TokenType::IF, "IF"),
//     (TokenType::ELSE, "ELSE"),
//     (TokenType::TRUE, "TRUE"),
//     (TokenType::FALSE, "FALSE"),
// ];

pub struct Token {
    pub tokentype: TokenType,
    pub literal: String,
}

impl Token {
    pub fn new(ttype: TokenType, literal: String) -> Self {
        Self {
            tokentype: ttype,
            literal: literal,
        }
    }
}
